clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala

```diff
@@ -0,0 +1,49 @@
+package org.broadinstitute.clio.client.dispatch
+
+import akka.http.scaladsl.model.HttpResponse
+import akka.http.scaladsl.model.headers.HttpCredentials
+import org.broadinstitute.clio.client.commands.{DeliverWgsCram, MoveWgsCram}
+import org.broadinstitute.clio.client.util.IoUtil
+import org.broadinstitute.clio.client.webclient.ClioWebClient
+import org.broadinstitute.clio.transfer.model.WgsCramIndex
+import org.broadinstitute.clio.transfer.model.wgscram.TransferWgsCramV1Metadata
+
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+  * Special-purpose CLP for delivering crams to Firecloud workspaces.
+  *
+  * Wraps the move CLP with extra IO / upsert logic.
+  */
+class DeliverExecutor(deliverCommand: DeliverWgsCram) extends Executor {
+
+  val moveCommand =
+    MoveWgsCram(deliverCommand.key, deliverCommand.workspacePath)
+  import moveCommand.index.implicits._
+
+  override def execute(webClient: ClioWebClient, ioUtil: IoUtil)(
+    implicit ex: ExecutionContext,
+    credentials: HttpCredentials
+  ): Future[HttpResponse] = {
+    import webClient.materializer
+
+    val moveExecutor = new MoveExecutor(moveCommand)
+
+    for {
+      moveResponse <- moveExecutor.execute(webClient, ioUtil)
+      _ = moveResponse.discardEntityBytes()
+      upsertResponse <- recordWorkspaceName(webClient)
+    } yield {
+      upsertResponse
+    }
+  }
+
+  private def recordWorkspaceName(
+    webClient: ClioWebClient
+  )(implicit credentials: HttpCredentials): Future[HttpResponse] = {
+    val newMetadata = TransferWgsCramV1Metadata(
+      workspaceName = Some(deliverCommand.workspaceName)
+    )
+    webClient.upsert(WgsCramIndex, deliverCommand.key, newMetadata)
```

Comment:
This seems odd in that we're just recording the value of `workspaceName` but not using it at all when storing the data. How do we know `workspaceName` corresponds to `workspacePath` at all? If we have `workspaceName`, can we derive `workspacePath`?

If we are delivering to a path in firecloud, how do we ensure we have permission to write to the firecloud bucket?

Reply:
In order:
1. We don't (as far as Clio is concerned)
2. We can't (asked Sam and Kathleen yesterday)
3. We don't (as far as Clio is concerned)

The delivery WDL that's going to use this CLP will be responsible for getting all these things right.

Reply:
OK